File: mobile-ios/TeacherWorkspace/Services/ScienceNotation.swift
import Foundation

/// 화면·공유 텍스트·PDF에 쓰기 직전 화학식과 전자배치를 글꼴 독립적인 유니코드 표기로 통일한다.
enum ScienceNotation {
    private static let subscripts = ["0": "₀", "1": "₁", "2": "₂", "3": "₃", "4": "₄", "5": "₅", "6": "₆", "7": "₇", "8": "₈", "9": "₉"]
    private static let superscripts = ["0": "⁰", "1": "¹", "2": "²", "3": "³", "4": "⁴", "5": "⁵", "6": "⁶", "7": "⁷", "8": "⁸", "9": "⁹", "+": "⁺", "-": "⁻"]

    static func format(_ value: String) -> String {
        var result = value.replacingOccurrences(of: "$", with: "")
        result = replace(result, pattern: #"\\text\{([^}]*)\}"#) { match, source in source.substring(with: match.range(at: 1)) }
        result = replace(result, pattern: #"\\delta\^\{?([+-])\}?"#) { match, source in source.substring(with: match.range(at: 1)) == "+" ? "δ⁺" : "δ⁻" }
        result = result.replacingOccurrences(of: "\\circ", with: "°")
        result = replace(result, pattern: #"_\{?([0-9]+)\}?"#) { match, source in toSubscript(source.substring(with: match.range(at: 1))) }
        result = replace(result, pattern: #"\^\{?([0-9+-]+)\}?"#) { match, source in toSuperscript(source.substring(with: match.range(at: 1))) }
        result = replace(result, pattern: #"\b(?:[A-Z][a-z]?\d*){2,}\b"#) { match, source in formatBareFormula(source.substring(with: match.range)) }
        result = replace(result, pattern: #"\b((?:[A-Z][a-z]?\d*)+)([+-])(?=\s|$|[),.?!])"#) { match, source in "\(formatBareFormula(source.substring(with: match.range(at: 1))))\(toSuperscript(source.substring(with: match.range(at: 2))))" }
        return replace(result, pattern: #"\b(\d+[spdfg])(\d+)\b"#) { match, source in "\(source.substring(with: match.range(at: 1)))\(toSuperscript(source.substring(with: match.range(at: 2))))" }
    }

    private static func formatBareFormula(_ formula: String) -> String {
        replace(formula, pattern: #"([A-Z][a-z]?)(\d+)"#) { match, source in "\(source.substring(with: match.range(at: 1)))\(toSubscript(source.substring(with: match.range(at: 2))))" }
    }

    private static func toSubscript(_ digits: String) -> String { digits.map { subscripts[String($0)] ?? String($0) }.joined() }
    private static func toSuperscript(_ characters: String) -> String { characters.map { superscripts[String($0)] ?? String($0) }.joined() }

    private static func replace(_ value: String, pattern: String, transform: (NSTextCheckingResult, NSString) -> String) -> String {
        guard let expression = try? NSRegularExpression(pattern: pattern) else { return value }
        let source = value as NSString
        let matches = expression.matches(in: value, range: NSRange(location: 0, length: source.length))
        return matches.reversed().reduce(value) { current, match in
            guard let range = Range(match.range, in: current) else { return current }
            return current.replacingCharacters(in: range, with: transform(match, source))
        }
    }
}

File: mobile-android/app/src/main/java/com/eunmastudio/teacherworkspace/ScienceNotation.kt
package com.eunmastudio.teacherworkspace

/** 화면·공유·PDF·DOCX를 만들기 직전에 화학식과 전자배치를 유니코드 아래첨자·위첨자로 통일한다. */
object ScienceNotation {
    private val subscript = mapOf('0' to '₀', '1' to '₁', '2' to '₂', '3' to '₃', '4' to '₄', '5' to '₅', '6' to '₆', '7' to '₇', '8' to '₈', '9' to '₉')
    private val superscript = mapOf('0' to '⁰', '1' to '¹', '2' to '²', '3' to '³', '4' to '⁴', '5' to '⁵', '6' to '⁶', '7' to '⁷', '8' to '⁸', '9' to '⁹', '+' to '⁺', '-' to '⁻')

    fun format(value: String): String {
        var text = value.replace(Regex("\\$([^$]+)\\$"), "$1")
            .replace(Regex("\\\\text\\{([^}]*)}"), "$1")
            .replace(Regex("\\\\delta\\^\\{?([+-])\\}?")) { if (it.groupValues[1] == "+") "δ⁺" else "δ⁻" }
            .replace("\\circ", "°")
            .replace(Regex("_\\{?([0-9]+)\\}?")) { toSubscript(it.groupValues[1]) }
            .replace(Regex("\\^\\{?([0-9+-]+)\\}?")) { toSuperscript(it.groupValues[1]) }
        text = text.replace(Regex("\\b(?:[A-Z][a-z]?\\d*){2,}\\b")) { formatBareFormula(it.value) }
        text = text.replace(Regex("\\b((?:[A-Z][a-z]?\\d*)+)([+-])(?=\\s|$|[),.?!])")) { "${formatBareFormula(it.groupValues[1])}${toSuperscript(it.groupValues[2])}" }
        return text.replace(Regex("\\b(\\d+[spdfg])(\\d+)\\b")) { "${it.groupValues[1]}${toSuperscript(it.groupValues[2])}" }
    }

    private fun formatBareFormula(value: String): String = value.replace(Regex("([A-Z][a-z]?)(\\d+)")) { "${it.groupValues[1]}${toSubscript(it.groupValues[2])}" }
    private fun toSubscript(value: String): String = value.map { subscript[it] ?: it }.joinToString("")
    private fun toSuperscript(value: String): String = value.map { superscript[it] ?: it }.joinToString("")
}
